tla-pp/src/main/scala/at/forsyte/apalache/tla/pp/IntegerQuantificationOptimizer.scala

```diff
@@ -0,0 +1,88 @@
+package at.forsyte.apalache.tla.pp
+
+import at.forsyte.apalache.tla.lir._
+import at.forsyte.apalache.tla.lir.oper._
+import at.forsyte.apalache.tla.lir.transformations.{TlaExTransformation, TransformationTracker}
+import at.forsyte.apalache.tla.lir.values.{TlaInt, TlaIntSet, TlaNatSet}
+
+import com.google.inject.Singleton
+
+/**
+ * This transformation rewrites expressions of the form `(\A | \E) x \in S: P`, where S is one of `Int`, `Nat`, `a..b`,
+ * into `(\A | \E) x: Q` (unbounded quantification).
+ *
+ * This is the only place we support unbounded quantification, as we can leverage the type annotation of `x` (which
+ * should be IntT1()). We can then introduce a special rewriting rule for this quantification form.
+ *
+ * @author
+ *   Jure Kukovec
+ */
+
+@Singleton
+class IntegerQuantificationOptimizer(tracker: TransformationTracker) extends TlaExTransformation {
```

Comment:
What motivates the use the `@Singleton` annotation here instead of defining this as an `object`?

Reply:
I think I copied `TypeSubstitutor.scala` as a template. It's probably not needed. It definitely can't be an `object` though, as it's parameterized by `tracker`.

Reply:
h right. Makes sense. We might want to figure out what that singleton annotation is used for at some point!